### PR TITLE
Load monitor data 2

### PR DESCRIPTION
**Pull request by Tom-Willemsen**
Closes https://github.com/scipp/scippneutron/issues/139

When loading data from histogram mode monitors, gives access to the structures as they appear in the nexus file. This seems to make sense to me as it may contain facility-specific implementation details (like periods at ISIS).

When loading event data, uses the same code & data structures as event mode data from detectors.

### CODE REVIEW

File: python/src/scippneutron/file_loading/_monitor_data.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        if nexus.dataset_in_group(group.group, "event_index")[0]:
+            event_groups.append(group)
+
+    # If any monitors contain event-mode data, load it
+    if event_groups:
+        monitor_events = load_detector_data(event_groups, [], file_root, nexus, True)
+        if isinstance(loaded_data, sc.DataArray):
+            loaded_data.attrs["monitor_events"] = sc.Variable(value=monitor_events)
+        else:
+            loaded_data["monitor_events"] = sc.Variable(value=monitor_events)
```
Comment (SimonHeybrock) [inline python/src/scippneutron/file_loading/_monitor_data.py:46]: Not your change/addition, but I feel it is weird that the helper functions that load meta data (such as monitors) do the looping and then also handle whether things are added as dataset entries or attributes. Wouldn't it make more sense to loop in `_load_data` and simply *return* (a single array) in the helper function, letting the caller decide whether to add as dataset items or attributes?

File: python/src/scippneutron/file_loading/_monitor_data.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        monitor_name = group.path.split("/")[-1]
+        data = _load_data_from_histogram_mode_monitor(group, file_root, nexus)
+        if isinstance(loaded_data, sc.DataArray):
+            loaded_data.attrs[monitor_name] = sc.Variable(value=data)
+        else:
+            loaded_data[monitor_name] = sc.Variable(value=data)
+
+        # Look for event mode data structures in NXMonitor
+        if nexus.dataset_in_group(group.group, "event_index")[0]:
+            event_groups.append(group)
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply python/src/scippneutron/file_loading/_monitor_data.py:38]: Is this allowed by Nexus? What is the interpretation of this? A pre-histogrammed version of the events? Maybe we can ignore the histogram in this case (and print a warning)?

File: python/src/scippneutron/file_loading/_monitor_data.py (hunk 1 above)
Comment (Tom-Willemsen) [reply python/src/scippneutron/file_loading/_monitor_data.py:46]: I've made most of the metadata functions return `Dict`, the caller now chooses to add the dict items as attributes or dataset items. This wasn't possible for some of the functions (e.g. those that modify coords directly), I've left those functions as-is.

File: python/src/scippneutron/file_loading/_monitor_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply python/src/scippneutron/file_loading/_monitor_data.py:38]: > Is this allowed by Nexus?

It's not entirely clear to me from [the docs](https://manual.nexusformat.org/classes/base_classes/NXmonitor.html), but I think so? The only example file I have does this. I would assume it is a pre-histogrammed version of the same data (possibly with the histogramming done at hardware level)

> Maybe we can ignore the histogram in this case (and print a warning)?

I guess I would ask the question the other way around - what is the harm in loading both if both are present? The histogram structures are likely to be quite small compared to events if they exist? I would be worried about the case where an event structure exists but is empty/useless, where we would then ignore the corresponding potentially useful histogram?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply python/src/scippneutron/file_loading/_monitor_data.py]: > what is the harm in loading both if both are present

The harm is that we have to invent new names that may clash with other things in the file, and make it hard to find items in the file/object.

[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply python/src/scippneutron/file_loading/_monitor_data.py]: ok, I've changed it to only load events if event structures are present. I guess in future we may want a flag to force loading all the data from the nexus file, but for this issue I guess it's probably ok.

File: python/src/scippneutron/file_loading/_monitor_data.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def load_monitor_data(monitor_groups: List[Group], file_root: h5py.File,
+                      nexus: LoadFromNexus) -> Dict:
+    monitor_data = {}
+    for group in monitor_groups:
+        monitor_name = group.path.split("/")[-1]
+
+        # Look for event mode data structures in NXMonitor. Event-mode data takes
+        # precedence over histogram-mode-data if available.
```
Comment (SimonHeybrock) [inline python/src/scippneutron/file_loading/_monitor_data.py:28]: Print a warning of both is found, saying that histogram data is ignored?

**Comment by SimonHeybrock** [comment]
@Tom-Willemsen Can you fix the conflict and add a warning as suggested. Also add a mention in the release notes.

**Review by SimonHeybrock — APPROVED** [review]
👍

**Pull request merged**